File: scripts/verify-deploy-guard.js
const fs = require("node:fs");
const { execSync } = require("node:child_process");

function fail(message) {
  console.error(`Deploy guard failed: ${message}`);
  process.exit(1);
}

function readJson(filePath) {
  if (!fs.existsSync(filePath)) {
    fail(`${filePath} is missing`);
  }
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch (error) {
    fail(`${filePath} is not valid JSON: ${error.message}`);
  }
}

const railwayJson = readJson("railway.json");
const deploy = railwayJson.deploy || {};

const nixpacksTomlPath = "nixpacks.toml";
if (!fs.existsSync(nixpacksTomlPath)) {
  fail("nixpacks.toml is required to force the Python build provider on Railway");
}

const nixpacksToml = fs.readFileSync(nixpacksTomlPath, "utf8");
if (!/providers\s*=\s*\[[^\]]*['"]python['"][^\]]*\]/i.test(nixpacksToml)) {
  fail("nixpacks.toml must explicitly set providers = ['python']");
}

if (!String(deploy.startCommand || "").includes("python manage.py runserver")) {
  fail("railway.json deploy.startCommand must run Django via 'python manage.py runserver'");
}

if (String(deploy.restartPolicyType || "").trim() !== "ON_FAILURE") {
  fail("railway.json deploy.restartPolicyType must be 'ON_FAILURE'");
}

const forbiddenTrackedPaths = [
  "railway.toml",
  "Dockerfile",
  "deploy.config.json",
  ".stripe-installer/deploy-manifest.json",
  ".stripe-installer/stripe-manifest.json",
];

const trackedFiles = new Set(
  execSync("git ls-files", { encoding: "utf8" })
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
);

for (const path of forbiddenTrackedPaths) {
  if (trackedFiles.has(path) && fs.existsSync(path)) {
    fail(`${path} must not be tracked in git`);
  }
}

console.log("Deploy guard passed.");

File: scripts/verify_deploy_guard.py
import json
import pathlib
import subprocess
import sys

ROOT = pathlib.Path(__file__).resolve().parents[1]


def fail(message: str) -> None:
    print(f"Deploy guard failed: {message}")
    sys.exit(1)


def read_json(path: pathlib.Path):
    if not path.exists():
        fail(f"{path.name} is missing")
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:
        fail(f"{path.name} is not valid JSON: {exc}")


def git_tracked_files() -> set[str]:
    result = subprocess.run(
        ["git", "ls-files"],
        cwd=ROOT,
        capture_output=True,
        text=True,
        check=True,
    )
    return {line.strip() for line in result.stdout.splitlines() if line.strip()}


railway_json = read_json(ROOT / "railway.json")
deploy = railway_json.get("deploy", {})

nixpacks_path = ROOT / "nixpacks.toml"
if not nixpacks_path.exists():
    fail("nixpacks.toml is required")

nixpacks_content = nixpacks_path.read_text(encoding="utf-8")
if "providers = ['python']" not in nixpacks_content and 'providers = ["python"]' not in nixpacks_content:
    fail("nixpacks.toml must explicitly set providers = ['python']")

start_command = str(deploy.get("startCommand", "")).strip()
if "python manage.py runserver" not in start_command:
    fail("railway.json deploy.startCommand must run Django with python manage.py runserver")

if str(deploy.get("restartPolicyType", "")).strip() != "ON_FAILURE":
    fail("railway.json deploy.restartPolicyType must be ON_FAILURE")

forbidden_tracked_paths = {
    "server.js",
    "package.json",
    "package-lock.json",
    "src/db.js",
    "src/services/automation.js",
    "src/services/salesPackage.js",
    "src/services/validationService.js",
    "scripts/connect-railway-stripe.js",
    "scripts/predeploy-check.js",
    "scripts/process-email-jobs.js",
    "scripts/verify-deploy-guard.js",
}

tracked = git_tracked_files()

for rel_path in sorted(forbidden_tracked_paths):
    if rel_path in tracked and (ROOT / rel_path).exists():
        fail(f"{rel_path} must not be tracked in Django-only runtime")

print("Deploy guard passed.")
